src/Npgsql/NpgsqlConnector.cs

```diff
@@ -593,6 +593,14 @@ async Task RawOpen(NpgsqlTimeout timeout, bool async, CancellationToken cancella
                 ReadBuffer = new NpgsqlReadBuffer(this, _stream, _socket, Settings.ReadBufferSize, TextEncoding, RelaxedTextEncoding);
                 WriteBuffer = new NpgsqlWriteBuffer(this, _stream, _socket, Settings.WriteBufferSize, TextEncoding);
 
+                if (timeout.IsSet)
+                {
+                    var timeLeft = timeout.TimeLeft;
+                    if (timeLeft > TimeSpan.Zero)
+                        UserTimeout = (int)Math.Ceiling(timeout.TimeLeft.TotalSeconds);
```

Comment:
`UserTimeout` only holds current timeout. Buffer's timeout is set during connector's `ReadMessage`.
Also, the connector sets the timeout only for the `ReadBuffer`.

Reply:
> UserTimeout only holds current timeout. Buffer's timeout is set during connector's ReadMessage.

Right, but the first time that a message will be read (i.e. for authentication) the timeout will be properly set on the read buffer - just like the regular flow when executing a command (that's why the test passes). In fact, if we don't set UserTimeout, the test fails because ReadMessage sets the buffer's timeout from it (and it means "no timeout").

>  Also, the connector sets the timeout only for the ReadBuffer.

That's true, I'll set the write timeout here as well. The design is currently smelly, in that the read timeout is set automatically from NpgsqlConnector.UserTimeout (in ReadMessage), but the write timeout is set via a completely different mechanism. This is mainly because reading always goes through one place (ReadMessage), but writing can happen via various different places.

Even though it's a code smell, I propose to leave it as it is as it doesn't have an actual impact - we can always clean up later if we want.

Reply:
> Right, but the first time that a message will be read (i.e. for authentication) the timeout will be properly set on the read buffer - just like the regular flow when executing a command (that's why the test passes).

But it wouldn't be set for reading a response for the ssl request just below.

> Even though it's a code smell, I propose to leave it as it is as it doesn't have an actual impact - we can always clean up later if we want.

It's definitely something we should take another look. But I agree, that it can (and should) be done later.

Reply:
> But it wouldn't be set for reading a response for the ssl request just below.

Good catch. I've updated to apply the timeout both directly to the read buffer, and also to UserTimeout.